### DISCUSSION TITLE
How to read from a pipe?

### DISCUSSION THREAD

**Discussion by helderco** [agronholm/anyio#499] [Q&A]
Hi!

Sorry for the abstract/pseudo code, I'm looking for a way to read a stream from a pipe.

Let's say I have a sync process where I want to read from its stderr from async.

```python
import anyio
import subprocess
import typing

class Engine:
    fd: typing.TextIO

    def start(self):
        ...
        self._proc = subprocess.Popen([...], stderr=self.fd)


class Connection:
    def __init__(self, fd):
        self.engine = Engine(fd)

    async def __aenter__(self):
        await anyio.to_thread.run_sync(self.engine.start)
        return self
```

What can I do with anyio that I can give to `Popen` so that I can do something with what gets written to it?

```python
async def main():
    pread, pwrite = os.pipe()
    reader = SomeReader(pread)

    # pass along pwrite through to Popen
    async with Connection(pwrite) as spam:
        async for line in reader:
            process(line)
```

**Comment by agronholm** [comment]
There is no direct support for async pipes, but you have two options:

1. Use AnyIO's async subprocess support
2. Wrap a pipe's file descriptor with `AsyncFile`

Would either of these work for you?
